ompi/mca/coll/han/coll_han_gather.c

```diff
@@ -251,18 +251,16 @@ int mca_coll_han_gather_lg_task(void *task_args)
                                    &rgap);
         tmp_buf = (char *) malloc(rsize);
         tmp_rbuf = tmp_buf - rgap;
-        if (t->w_rank == t->root) {
-            if (MPI_IN_PLACE == t->sbuf) {
-                ptrdiff_t rextent;
-                ompi_datatype_type_extent(dtype, &rextent);
-                ptrdiff_t block_size = rextent * (ptrdiff_t)count;
-                ptrdiff_t src_shift = block_size * t->w_rank;
-                ptrdiff_t dest_shift = block_size * low_rank;
-                ompi_datatype_copy_content_same_ddt(dtype,
-                                                    (ptrdiff_t)count,
-                                                    tmp_rbuf + dest_shift,
-                                                    (char *)t->rbuf + src_shift);
-            }
+        if (t->w_rank == t->root && MPI_IN_PLACE == t->sbuf) {
```

Comment:
This code will not work if `t->w_rank == low_rank`, because we will force the convertor to do an inplace copy and the convertor uses memcpy not memmove. Please add `&& (t->w_rank != low_rank)` to your condition.

We could have also avoided this copy by pointing the `sbuf` to the right location (aka `t->rbuf + src_shift`) and let the gather do the copy, because this block is only executed on the root of the user collective (top level).

Reply:
I'm confused? Was this always broken? I didn't change anything except for combine the if checks

Reply:
My bad, it is working because opal_datatype_copy_content_same_ddt has a special check to see if it should use memcpy or memove (opal/datatype/opal_datatype_copy.c). Thus, it is working but suboptimal.